src/miniwdl_slurm/__init__.py

```diff
@@ -0,0 +1,115 @@
+
+from WDL import Type, Value
+from WDL.runtime import config
+from WDL.runtime.backend.cli_subprocess import _SubprocessScheduler
+from WDL.runtime.backend.singularity import SingularityContainer
+
+
+class SlurmSingularity(SingularityContainer):
+    @classmethod
+    def global_init(cls, cfg: config.Loader, logger: logging.Logger) -> None:
+        # Set resources to maxsize. The base class (_SubProcessScheduler)
+        # looks at the resources of the current host, but since we are
+        # dealing with a cluster these limits do not apply.
+        cls._resource_limits = {
+            "cpu": sys.maxsize,
+            "mem_bytes": sys.maxsize,
+            "time": sys.maxsize,
+        }
+        _SubprocessScheduler.global_init(cls._resource_limits)
+        # Since we run on the cluster, the images need to be placed in a
+        # shared directory. The singularity cache itself cannot be shared
+        # across nodes, as it can become corrupted when nodes pull the same
+        # image. The solution is to pull image to a shared directory on the
+        # submit node. If no image_cache is given, simply place a folder in
+        # the current working directory.
+        if cfg.get("singularity", "image_cache") == "":
+            cfg.override(
+                {"singularity": {
+                    "image_cache": os.path.join(os.getcwd(),
+                                                "miniwdl_singularity_cache")
+                }}
+            )
+        SingularityContainer.global_init(cfg, logger)
+
+    @classmethod
+    def detect_resource_limits(cls, cfg: config.Loader,
+                               logger: logging.Logger) -> Dict[str, int]:
+        return cls._resource_limits  # type: ignore
+
+    @property
+    def cli_name(self) -> str:
+        return "slurm_singularity"
+
+    def process_runtime(self,
+                        logger: logging.Logger,
+                        runtime_eval: Dict[str, Value.Base]) -> None:
+        """Any non-default runtime variables can be parsed here"""
+        super().process_runtime(logger, runtime_eval)
+        if "time_minutes" in runtime_eval:
+            time_minutes = runtime_eval["time_minutes"].coerce(Type.Int()).value
+            self.runtime_values["time_minutes"] = time_minutes
+
+    def _slurm_invocation(self):
+        # We use srun as this makes the submitted job behave like a local job.
+        # This also gives informative exit codes back, including 253 for out
+        # of memory.
+        srun_args = [
+            "srun",
+            "--job-name", self.run_id,
+        ]
+
+        cpu = self.runtime_values.get("cpu", None)
+        if cpu is not None:
+            srun_args.extend(["--cpus-per-task", str(cpu)])
+
+        memory = self.runtime_values.get("memory_reservation", None)
+        if memory is not None:
+            # Round to the nearest megabyte.
+            srun_args.extend(["--mem", f"{round(memory / (1024 ** 2))}M"])
```

Comment:
Uhm actually 🤓, 1024**2 is technically a mebibyte, not a megabyte. I wonder if wdl/miniwdl and slurm consider `M` to mean `MB` or `MiB`. If they consider it to mean `MB` then I think this would lead to less memory being requested than was asked for.

E.g. if the WDL states:
```wdl
runtime { memory: "100M" }
``` 
and miniwdl an slurm interprets that as MB, then this would lead to `100,000,000 / 1,048,576 ≈ 95`. So, 95 MB would get requested from slurm, instead of the intended 100 MB.

Reply:
Slurm utilizes mebibytes internally: https://github.com/SchedMD/slurm/blob/95bdcfc1daccf917bceac83f37662193c3cfc1f7/src/plugins/task/cgroup/task_cgroup_memory.c#L157